src/Microsoft.Azure.WebJobs.Extensions.ServiceBus/Listeners/ServiceBusListener.cs

```diff
@@ -46,8 +59,27 @@ public Task StartAsync(CancellationToken cancellationToken)
                 throw new InvalidOperationException("The listener has already been started.");
             }
 
-            _receiver = _messagingProvider.CreateMessageReceiver(_entityPath, _serviceBusAccount.ConnectionString);
-            _receiver.RegisterMessageHandler(ProcessMessageAsync, _messageProcessor.MessageOptions);
+            if (_entity.IsSesssionsEnabled)
+            {
+                _clientEntity = _messagingProvider.CreateClientEntity(_entity.EntityPath, _serviceBusAccount.ConnectionString);
+
+                // https://github.com/Azure/azure-service-bus/blob/73aaa2377e8fdff02d2932230741d44b72d24a92/samples/DotNet/Microsoft.Azure.ServiceBus/Sessions/Program.cs#L85
+                QueueClient queueClient = _clientEntity as QueueClient;
+                if (queueClient != null)
+                {
+                    queueClient.RegisterSessionHandler(ProcessSessionMessageAsync, _serviceBusOptions.SessionHandlerOptions);
+                }
+                else
+                {
+                    SubscriptionClient subscriptionClient = _clientEntity as SubscriptionClient;
+                    subscriptionClient.RegisterSessionHandler(ProcessSessionMessageAsync, _serviceBusOptions.SessionHandlerOptions);
+                }
+            }
+            else
+            {
+                _receiver = _messagingProvider.CreateMessageReceiver(_entity.EntityPath, _serviceBusAccount.ConnectionString);
```

Comment:
Technically we could be going through the new CreateClientEntity path for non-sessions right? We could create a Queue/Subscription Client and call RegisterMessageHandler. I get that we have a back compat issue here though.

Reply:
Technically yes. I left MessageReceiver.RegisterMessageHandler for back compatibility.